### PR TITLE
Add destructors and fix up tests

### PR DESCRIPTION
**Pull request by atgeller**

### CODE REVIEW

File: lib/Optimizer/Transforms/DependencyAnalysis.cpp [hunk 1]
```diff
@@ -130,6 +130,7 @@ class LifeTimeAnalysis {
     if (best_reuse) {
       auto physical = best_reuse.value();
       lifetimes[physical]->combine(lifetime);
+      delete lifetime;
```
Comment (bmhowe23) [inline lib/Optimizer/Transforms/DependencyAnalysis.cpp:133]: Is it safe to conditionally delete a pointer in this function? How will the caller know if it was deleted or not?
Semi-related question: does it also need to be deleted after line 140?

Comment (atgeller) [reply lib/Optimizer/Transforms/DependencyAnalysis.cpp:133]: I think this is a side effect of the fact that the lifetimes here probably shouldn't be pointers, which will simplify away these issues.

**Review by bmhowe23 — APPROVED** [review]
👍

**Pull request merged**